TVB-2930 + TVB-2931 : Solved project deletion issues

VinczeRobert opened the pull request:
I solved both problems because they were related.

Project deletion with links until now only worked if we linked a datatype that is not linked to some other datatypes. If we had, for example, a region mapping then it didn't copy the connectivity and the surface to the new project, thus we couldn't view the region mapping.

However, the bigger problem was that copying linked datatype groups didn't work at all. I managed to solve this as well. Now it doesn't matter if we link the datatype group for TSIndex or DMIndex, in both cases both of them will be copied to the new project.

If we have a datatype that is linked to more projects then it will be moved to the first project returned by the database and it will have a link pointing to the other projects.

I have tested several cases both in Sqlite and Postgres and didn't find any issues.
Maybe the project deletion process is too complicated and we talked with Paula about reviewing generally the remove methods. I intend to do that soon but for now, I think this is okay to be merged.

liadomide commented on framework_tvb/tvb/core/services/project_service.py:
```diff
@@ -295,16 +298,23 @@ def remove_project(self, project_id):
+            # a Region Mapping and a Time Series
+            projects = dao.get_all_projects()
+            for project in projects:
+                datatypes = dao.get_datatypes_in_project(project.id)
```
Waw. These queries are extremely costly. 
Why not to keep (maybe fix) the original intended query ?

VinczeRobert replied:
Because that query was deleting the links from the project to be deleted and after I refactored the code by the time it would reach that line, those links are already deleted. My queries are intending to remove redundant links: links which point to the same project that the datatype is in. I will think of doing it without these 3 fors...

liadomide commented on framework_tvb/tvb/core/services/project_service.py:
```diff
@@ -36,10 +36,13 @@
 import formencode
+import os
 
+from tvb.adapters.datatypes.db.time_series import TimeSeriesIndex
```
This dependency is not good here (from 'core' towards 'adapters')

VinczeRobert replied:
True, I can check some attribute here that time series have and datatype measures don't, like 'data_ndim'.

VinczeRobert commented:
I removed those three fors now by keeping in a list the additional links that need to be created and this way we don't add the same link twice. 

Also, I don't understand why when we link a datatype group we are making a links towards each data of the group. I changed it to add the link only for the group and I don't see any errors, but if there is a reason, please enligthen me.

The only problem with this solution is that there are two cases when redundant links (links that point towards the same project that the datatype is in) will remain in the database, but they will not cause any errors in the functionality:
1. if the user links both the ts group and the datatype measure group, the link for the ts group will remain
2. e.g. if the user links a region mapping and the connectivity that was used for the region mapping: a link towards the conn will remain.
These 2 problems can be solved by additional conditions but that would just further complicate the solution. Truth is, its pretty hard to copy all kinds of linked datatypes when a project is deleted.

popaula937 commented on framework_tvb/tvb/core/services/project_service.py:
```diff
@@ -521,65 +523,123 @@ def get_datatype_details(datatype_gid):
+
+            dt = h5.load(h5_path)
+            dt_index = h5.load_entity_by_gid(dt.gid)
+            new_link = Links(dt_index.id, fk_to_project)
```
Could we avoid loading the `dt` and `dt_index` here, and find the gid directly from the H5 file with the `H5File.get_metadata_param` method which is static?

VinczeRobert replied:
But I need the id, not the gid.

popaula937 commented on framework_tvb/tvb/core/services/project_service.py:
```diff
@@ -521,65 +523,123 @@ def get_datatype_details(datatype_gid):
+
+                # If there is a  datatype group and operation group and they were not moved yet to the linked project,
+                # then do it
+                if datatype.fk_datatype_group is not None:
```
isn't this called for every dt in a group? Could we avoid that?

VinczeRobert replied:
No it is not. The condition you asked before at line 673 makes the flow not reach this method for every datatype in the group.

VinczeRobert commented on framework_tvb/tvb/core/services/project_service.py:
```diff
@@ -599,7 +659,7 @@ def remove_operation(self, operation_id):
             self.logger.warning("Attempt to delete operation with id=%s which no longer exists." % operation_id)
 
-    def remove_datatype(self, project_id, datatype_gid, skip_validation=False):
+    def remove_datatype(self, project_id, datatype_gid, skip_validation=False, existing_dt_links=[]):
```
I saw this in other places this as well but this one can be easily changed to None so it is okay.

VinczeRobert commented on framework_tvb/tvb/core/services/project_service.py:
```diff
@@ -610,11 +670,20 @@ def remove_datatype(self, project_id, datatype_gid, skip_validation=False):
             self.logger.warning("Attempt to delete DT[%s] which no longer exists." % datatype_gid)
             return
 
+        if datatype.parent_operation.fk_launched_in != int(project_id):
```
When a project is deleted which has a datatype group being linked to another project. When the first member of the group will be removed, all the related files (all TSs and Datatypemeasures) are moved to the new project thus the return statement.

VinczeRobert commented on framework_tvb/tvb/core/services/project_service.py:
```diff
@@ -521,65 +523,123 @@ def get_datatype_details(datatype_gid):
 
-    def _remove_project_node_files(self, project_id, gid, skip_validation=False):
+    @staticmethod
+    def _add_links_for_linked_datatypes(datatype, fk_to_project, link_to_delete, existing_dt_links):
```
Of course.
